chanbackup, server, rpcserver: put close unsigned tx, remote signature and commit height to SCB

starius opened the pull request:
## Change Description

In this PR I'm implementing the proposal from https://github.com/lightningnetwork/lnd/issues/7658#issuecomment-1774983789

I changed `chanbackup` to add `CloseTxInputs` optional field to `chanbackup.Single` structure. When the field is present, the version byte is XORed with `closeTxVersion = 128`. When this bit is present in the version byte, the field is properly extracted when decoding packaged single backup.

I added `chantools scbforceclose` command to extract and sign closing tx. https://github.com/lightninglabs/chantools/pull/95

Implemented:
 * [x] add global and maybe also lncli flags enabling close tx in SCB (now it is always on)
   * [x] documentation for the flag and the feature
 * [x] tests for new features of chanbackup
 * [x] integration tests
 * [x] lines wrap at 80
 * [x] commit structure
 * [x] when LND shuts down, update channel.backup with latest close tx inputs

## Steps to Test

I tested the following scenario on testnet:
 * `lncli openchannel`
 * `lncli exportchanbackup --chan_point xxx:x --output_file single.backup`
 * `lncli stop`
 * `chantools --testnet scbforceclose --single_backup xxx`
 * Manually broadcasted the tx
 * Started LND from scratch on another machine
 * `lncli create`. Use existing seed, recover onchain funds.
 * `lncli restorechanbackup --single_file single.backup`
 * find the channel in `lncli pendingchannels`, `total_limbo_balance` is equal to channel balance minus fees

## Pull Request Checklist
### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Enhanced backup configuration flexibility with `BackupOption` parameters.
	- Added support for force close transaction data in backups.
	- Updated backup commands to include an option for force close transaction inputs.
	- Introduced a new option in LND for including data for unilateral channel closure in an SCB file.
	- Added a new parameter for controlling backup behavior in channel restoration scenarios.
	- Included an option to specify inclusion of inputs of the latest force close transaction in channel backups.
- **Bug Fixes**
	- Fixed a typo in the log message within the `SetupStandbyNodes` function.
- **Refactor**
	- Refined backup creation and serialization/deserialization logic to accommodate new backup options.
	- Updated `LightningChannel` to use a more modular approach for signing commitment transactions.
- **Documentation**
	- Improved explanations on commitment transaction sweeping in various scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lightninglabs-deploy commented:
@guggero: review reminder

guggero commented on chanbackup/backup.go:
```diff
@@ -53,11 +56,44 @@ func assembleChanBackup(addrSource AddressSource,
+	if err != nil {
+		return nil, fmt.Errorf("lnwallet.NewLightningChannel failed: %w", err)
+	}
+	closeSummary, err := chanMachine.GetLocalForceCloseSummary()
```
This will return a signed force close transaction. But I think for security reasons we should NOT have the signed TX in the SCB. Because that way it would be too easy for someone to decode the SCB, take the TX and publish it, breaching themselves.
I think we should only make it possible to extract and sign the TX using the appropriate `chantools` command, where we can show all the warnings and ask for confirmation several times...

starius replied:
I see two possible approaches how to address the security concerns.

1. Store two separate fields in SCB: unsigned transaction and encrypted signature of it. In this case the transaction is unlikely to be extracted from SCB and published accidentally, because it has to be decrypted and reassembled.
2. Actually store in SCB all the inputs except the signer. To do that, the logic of signing should be factored out from the code of [LightningChannel.getSignedCommitTx](https://github.com/lightningnetwork/lnd/blob/v0.17.3-beta/lnwallet/channel.go#L6249-L6358) and missing inputs should be added to SCB. Fortunately, some of them are already in SCB or are easy to add. The most problematic input is [input.SignDescriptor](https://pkg.go.dev/github.com/lightningnetwork/lnd@v0.17.3-beta/input#SignDescriptor) which has interface field PrevOutputFetcher. Encode and decode code is needed for these types to store them in SCB. If new channel type appears, new variables will be needed in SCB to produce commit tx.

I'm inclining to the first approach, because it is sufficient to improve security of SCB with regard to accidental transaction publishing, but it is isolated in the code base and does not add maintenance burden for future channel types.

What do you think?

starius replied:
To create a signed force close transaction, the following additional information is needed in SCB:

 * CommitTx (unsigned)
 * CommitSig
 * CommitHeight (for taproot channels only)

All other inputs seem to be already in SCB file.

I'll test how it works altogether in lnd & chantools.

starius replied:
I implemented the changes for the second approach. It works! I produced valid transactions in testnet using `chantools scbforceclose`.

guggero commented on chanbackup/backup.go:
```diff
@@ -53,11 +56,44 @@ func assembleChanBackup(addrSource AddressSource,
+
+type BackupOption func(*BackupConfig)
+
+func WithCloseTx(signer input.Signer) BackupOption {
```
Though we might not need the signer at all if we don't include a signed version of the TX.

starius replied:
In the recent version I still use an option to configure if this data should be included into a backup. For now I just set it always on in server.go and rpcserver.go, but I think it should be opt-in when released, so should be configurable.

guggero reviewed (commented):
Took a first look. I think the general idea is nice, this should help in some extreme edge cases.
Though the way we currently update those channels the commitment TX within the backup would be outdated very quickly for normal channels. So as a safety precaution we should not include a signature to make it hard for people to accidentally breach themselves.

lightninglabs-deploy commented:
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@starius, remember to re-request review from reviewers when ready

starius commented on chanbackup/single.go:
```diff
@@ -52,6 +52,9 @@ const (
 	SimpleTaprootVersion = 5
+
+	// XORed to some version on wire indicating that the backup has CloseTx.
+	closeTxVersion = 8
```
Fixed. I used 1 << 7, which is equal to 128.

starius commented on chanbackup/single.go:
```diff
@@ -327,9 +332,18 @@ func (s *Single) Serialize(w io.Writer) error {
+	version := s.Version
+	if len(s.CloseTx) != 0 {
+		version |= closeTxVersion
+		err := lnwire.WriteElements(&singleBytes, uint16(len(s.CloseTx)), s.CloseTx)
```
Fixed

starius commented on contractcourt/commit_sweep_resolver.go:
```diff
@@ -25,9 +25,7 @@ const (
-// output paying to us, in the case that the remote party broadcasts their
-// version of the commitment transaction. We can sweep this output immediately,
-// as it doesn't have a time-lock delay.
+// output paying to us.
```
Fixed. Added information about delays, anchor channels and leased channels.

starius commented on chanbackup/backup.go:
```diff
@@ -53,11 +56,44 @@ func assembleChanBackup(addrSource AddressSource,
 }
 
+func buildCloseTx(targetChan *channeldb.OpenChannel, signer input.Signer) ([]byte, error) {
+	chanMachine, err := lnwallet.NewLightningChannel(signer, targetChan, nil)
```
Fixed. Passed `make lint`.

starius commented:
@guggero Thank you for feedback! Happy New Year!

I updated both PRs and also their descriptions. SCB file now contains unsigned CommitTx, CommitSig of remote party, and for taproot channels also CommitHeight. I kept unresolved the comments where further acctention may be needed.

Please take another look!

starius commented:
I fixed few bugs in this PR.

1. Integration test [failed](https://app.travis-ci.com/github/lightningnetwork/lnd/jobs/615662242). I reproduced the failure using the following command (bitcoind binary is needed in PATH):

```
make itest icase=channel_backup_restore_basic backend=bitcoind
```

I found the root cause in log file `itest/.logs-tranche0/4-channel_backup_restore_basic-dave-021bb9a5.log` .

<details>
  <summary>panic stack trace</summary>
<pre>
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x21ec98]

goroutine 1 [running]:
github.com/btcsuite/btcd/wire.(*MsgTx).BtcEncode(0x0, {0x14691a0, 0x4001c3f500?}, 0x26a2832?, 0x2)
        /root/go/pkg/mod/github.com/btcsuite/btcd@v0.23.5-0.20230905170901-80f5a0ffdf36/wire/msgtx.go:736 +0x38
github.com/btcsuite/btcd/wire.(*MsgTx).Serialize(...)
        /root/go/pkg/mod/github.com/btcsuite/btcd@v0.23.5-0.20230905170901-80f5a0ffdf36/wire/msgtx.go:828
github.com/lightningnetwork/lnd/chanbackup.(*Single).Serialize(0x40004141a0, {0x14691a0?, 0x4001c3f470})
        /root/lnd/chanbackup/single.go:363 +0x560
github.com/lightningnetwork/lnd/chanbackup.Multi.PackToWriter({0xa8?, {0x4000414000?, 0x400000d110?, 0x4000169f88?}}, {0x14691a0, 0x4001c3f440}, {0xffff6c95df30, 0x400000d110})
        /root/lnd/chanbackup/multi.go:84 +0x240
github.com/lightningnetwork/lnd/chanbackup.(*SubSwapper).updateBackupFile(0x40001e40e0, {0x0, 0x0, 0x0?})
        /root/lnd/chanbackup/pubsub.go:238 +0x48c
github.com/lightningnetwork/lnd/chanbackup.(*SubSwapper).Start.func1()
        /root/lnd/chanbackup/pubsub.go:157 +0x6c
sync.(*Once).doSlow(0x4000716c60?, 0x0?)
        /usr/local/go/src/sync/once.go:74 +0x100
sync.(*Once).Do(...)
        /usr/local/go/src/sync/once.go:65
github.com/lightningnetwork/lnd/chanbackup.(*SubSwapper).Start(0x400049a420?)
        /root/lnd/chanbackup/pubsub.go:151 +0x50
github.com/lightningnetwork/lnd.(*server).Start.func1()
        /root/lnd/server.go:2072 +0x2120
sync.(*Once).doSlow(0x4?, 0x11b931e?)
        /usr/local/go/src/sync/once.go:74 +0x100
sync.(*Once).Do(...)
        /usr/local/go/src/sync/once.go:65
github.com/lightningnetwork/lnd.(*server).Start(0x1bf?)
        /root/lnd/server.go:1864 +0x6c
github.com/lightningnetwork/lnd.Main(0x400002ab00, {{0x0?, 0x40001822a0?, 0x4000182300?}}, 0x400010f080, {0x40001a2480, 0x40001822a0, 0x4000182300, 0x40001823c0, {0x0}})
        /root/lnd/lnd.go:684 +0x3164
main.main()
        /root/lnd/cmd/lnd/main.go:38 +0x198
</pre>
</details>

LND panicked trying to serialize a backup with nil CommitTx inside non-nil CloseTxInputs. This is [possible](https://github.com/lightningnetwork/lnd/blob/4d8fa349ca405a42548717c17cc689e6612ff37a/rpcserver.go#L3729) when DLP is active and this is exactly what is tested in the failing test (channel_backup_restore_basic).

I changed buildCloseTxInputs to return nil `*CloseTxInputs` producing classic SCB file in this case.

2. Also I found a mistake in error handling in Single.Serialize, returning `nil` instead of error. Fixed it.

3. Also I figured out that wire.MsgTx can be serialized directly to io.Writer, without a temp buffer. Putting length of its serialization is redundant. I changed it to read and write tx directly.

lightninglabs-deploy commented:
@guggero: review reminder

lightninglabs-deploy commented:
@guggero: review reminder

lightninglabs-deploy commented:
@guggero: review reminder

lightninglabs-deploy commented:
@guggero: review reminder

lightninglabs-deploy commented:
@guggero: review reminder

lightninglabs-deploy commented:
@guggero: review reminder

Roasbeef commented:
@coderabbitai review

lightninglabs-deploy commented:
@guggero: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero: review reminder
@starius, remember to re-request review from reviewers when ready

starius commented on rpcserver.go:
```diff
@@ -7191,6 +7191,7 @@ func (r *rpcServer) ExportChannelBackup(ctx context.Context,
 	// unknown, then we'll return an error
 	unpackedBackup, err := chanbackup.FetchBackupForChan(
 		chanPoint, r.server.chanStateDB, r.server.addrSource,
+		chanbackup.WithCloseTxInputs(true),
```
Fixed. Added flags `lncli exportchanbackup --with_close_tx_inputs` and `lnd --backupclosetxinputs` which enable inclusion of `CloseTxInputs` in backups. Added documentation about it to docs/recovery.md.

starius commented on rpcserver.go:
```diff
@@ -7361,6 +7362,7 @@ func (r *rpcServer) ExportAllChannelBackups(ctx context.Context,
 	// channels from disk.
 	allUnpackedBackups, err := chanbackup.FetchStaticChanBackups(
 		r.server.chanStateDB, r.server.addrSource,
+		chanbackup.WithCloseTxInputs(true),
```
Fixed. See my response to the previous comment.

starius commented on rpcserver.go:
```diff
@@ -7490,6 +7492,7 @@ func (r *rpcServer) SubscribeChannelBackups(req *lnrpc.ChannelBackupSubscription
 			// backups from disk.
 			chanBackups, err := chanbackup.FetchStaticChanBackups(
 				r.server.chanStateDB, r.server.addrSource,
+				chanbackup.WithCloseTxInputs(true),
```
Fixed. See my response to the previous comment.

starius commented on chanbackup/backup.go:
```diff
@@ -53,11 +53,60 @@ func assembleChanBackup(addrSource AddressSource,
+	return inputs
+}
+
+// BackupConfig contains options of backup creation process.
+type BackupConfig struct {
+	// Whether to put CloseTxInputs into a backup. A backup with this data
+	// can be used by "chantools scbforceclose" command.
+	includeCloseTxInputs bool
+}
```
The field is already documented. I think this comment is not relevant.

starius commented:
I addressed feedback from @coderabbitai

  - added options to API requests and LND option to explicitly enable inclusion of CloseTxInputs into backups
  - documented new options
  - changed itest/lnd_channel_backup_test to run the tests in both modes (with the feature enabled/disabled via `backupclosetxinputs` flag of LND)

starius commented on lnwallet/channel.go:
```diff
@@ -6458,29 +6458,68 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
+	TaprootNonceProducer shachain.Producer
+}
+
+// GetSignedCommitTx function takes the latest commitment transaction and
+// populates it with witness data. This pure function is needed to be able
+// to store the inputs without a signature in SCB and sign the transaction
+// in chantools scbforceclose.
+func GetSignedCommitTx(inputs SignedCommitTxInputs,
+	signer input.Signer) (*wire.MsgTx, error) {
```
Yes, please. Could you propose a test for this function, please? You can make two subtests: for regular channel and for taproot channel.

starius replied:
The structure looks good to me. Please proceed with implementing this test.

starius replied:
Proceed with general cases that cover a broad range of possibilities, please.

starius replied:
You can use a real signer with a random key instead of a mock signer.

starius commented on lnwallet/channel.go:
```diff
@@ -6458,29 +6458,68 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
+	commitTx := inputs.CommitTx.Copy()
 
 	var witness wire.TxWitness
 	switch {
 	// If this is a taproot channel, then we'll need to re-derive the nonce
 	// we need to generate a new signature
-	case lc.channelState.ChanType.IsTaproot():
+	case inputs.Taproot != nil:
 		// First, we'll need to re-derive the local nonce we sent to
 		// the remote party to create this musig session. We pass in
 		// the same height here as we're generating the nonce needed
 		// for the _current_ state.
 		localNonce, err := channeldb.NewMusigVerificationNonce(
-			ourKey.PubKey, lc.currentHeight,
-			lc.taprootNonceProducer,
+			inputs.OurKey.PubKey, inputs.Taproot.CommitHeight,
+			inputs.Taproot.TaprootNonceProducer,
```
This is covered by the above discussion.

starius commented on lnwallet/channel.go:
```diff
@@ -6490,13 +6529,13 @@ func (lc *LightningChannel) getSignedCommitTx() (*wire.MsgTx, error) {
 		musigSession := NewPartialMusigSession(
-			*localNonce, ourKey, theirKey, lc.Signer,
-			&lc.fundingOutput, LocalMusigCommit,
+			*localNonce, inputs.OurKey, inputs.TheirKey, signer,
+			inputs.SignDesc.Output, LocalMusigCommit,
```
This is covered by the above discussion.

starius commented on lnwallet/channel.go:
```diff
@@ -6542,25 +6581,25 @@ func (lc *LightningChannel) getSignedCommitTx() (*wire.MsgTx, error) {
 			return nil, err
 		}
 
 		// With this, we then generate the full witness so the caller
 		// can broadcast a fully signed transaction.
-		lc.signDesc.SigHashes = input.NewTxSigHashesV0Only(commitTx)
-		ourSig, err := lc.Signer.SignOutputRaw(commitTx, lc.signDesc)
+		inputs.SignDesc.SigHashes = input.NewTxSigHashesV0Only(commitTx)
+		ourSig, err := signer.SignOutputRaw(commitTx, inputs.SignDesc)
 		if err != nil {
 			return nil, err
 		}
 
 		// With the final signature generated, create the witness stack
 		// required to spend from the multi-sig output.
 		witness = input.SpendMultiSig(
-			lc.signDesc.WitnessScript,
-			ourKey.PubKey.SerializeCompressed(), ourSig,
-			theirKey.PubKey.SerializeCompressed(), theirSig,
+			inputs.SignDesc.WitnessScript,
+			inputs.OurKey.PubKey.SerializeCompressed(), ourSig,
+			inputs.TheirKey.PubKey.SerializeCompressed(), theirSig,
```
This is covered by the above discussion.

starius commented on lnwallet/channel.go:
```diff
@@ -6569,6 +6608,31 @@ func (lc *LightningChannel) getSignedCommitTx() (*wire.MsgTx, error) {
 	return commitTx, nil
 }
 
+// getSignedCommitTx function takes the latest commitment transaction and
+// populates it with witness data.
+func (lc *LightningChannel) getSignedCommitTx() (*wire.MsgTx, error) {
+	// Fetch the current commitment transaction, along with their signature
+	// for the transaction.
+	localCommit := lc.channelState.LocalCommitment
+
+	inputs := SignedCommitTxInputs{
+		CommitTx:  localCommit.CommitTx,
+		CommitSig: localCommit.CommitSig,
+		OurKey:    lc.channelState.LocalChanCfg.MultiSigKey,
+		TheirKey:  lc.channelState.RemoteChanCfg.MultiSigKey,
+		SignDesc:  lc.signDesc,
+	}
+
+	if lc.channelState.ChanType.IsTaproot() {
+		inputs.Taproot = &TaprootSignedCommitTxInputs{
+			CommitHeight:         lc.currentHeight,
+			TaprootNonceProducer: lc.taprootNonceProducer,
+		}
+	}
+
+	return GetSignedCommitTx(inputs, lc.Signer)
+}
```
No, let's skip creating unit tests for this method. It is already well tested by existing tests.

starius commented on server.go:
```diff
@@ -4656,6 +4664,39 @@ func (s *server) applyChannelUpdate(update *lnwire.ChannelUpdate,
 	}
 }
 
+// updateChannelBackup reads channels from DB, creates channel backup for all
+// the channels and writes it to disk.
+func (s *server) updateChannelBackup() error {
+	// s.chanStateDB is still active in server.Stop(), it is closed outside
+	// of server type, in function Main.
+	chans, err := chanbackup.FetchStaticChanBackups(
+		s.chanStateDB, s.addrSource,
+		chanbackup.WithCloseTxInputs(s.cfg.BackupCloseTxInputs),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to fetch channel backups: %w", err)
+	}
+
+	backupFile := chanbackup.NewMultiFile(s.cfg.BackupFilePath)
+
+	multi := chanbackup.Multi{
+		Version:       chanbackup.DefaultMultiVersion,
+		StaticBackups: chans,
+	}
+
+	var b bytes.Buffer
+	if err := multi.PackToWriter(&b, s.cc.KeyRing); err != nil {
+		return fmt.Errorf("unable to pack multi backup: %w", err)
+	}
+
+	packed := chanbackup.PackedMulti(b.Bytes())
+	if err := backupFile.UpdateAndSwap(packed); err != nil {
+		return fmt.Errorf("unable to update multi backup: %w", err)
+	}
+
+	return nil
+}
```
This is a helper function called when the server is shutting down to update the backup file the final time. It should not be customizable with version or filepath.

starius replied:
Actions in this function should be commented. Can you propose a change with that, please?

starius replied:
Thanks! I partially applied this.

starius commented:
There are no tests that the backup with CloseTxInputs can be used to produce a commit transaction. To add such tests, the code signing commit tx should be available to such a test. Currently the code is in https://github.com/lightninglabs/chantools/pull/95 in file `cmd/chantools/scbforceclose.go` in function `signCloseTx`. One approach is to move the function to LND and test it here, unit tests and itest. Another approach is to make an optional itest depending on `chantools` which runs `chantools scbforceclose` command. Any ideas on should this be tested in the first place and which of the options is preferred?

lightninglabs-deploy commented:
@guggero: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero: review reminder
@starius, remember to re-request review from reviewers when ready

guggero commented:
@saubyk this is a nice security/recovery feature. Should we prioritize it for v0.19?

lightninglabs-deploy commented:
@starius, remember to re-request review from reviewers when ready

saubyk commented:
> @saubyk this is a nice security/recovery feature. Should we prioritize it for v0.19?

Tagged for 0.19

lightninglabs-deploy commented:
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@starius, remember to re-request review from reviewers when ready

starius commented:
Rebased on top of 0.18, added release notes, updated PR description.
Retested manually using [chantools scbforceclose](https://github.com/lightninglabs/chantools/pull/95).

lightninglabs-deploy commented:
@guggero: review reminder
@starius, remember to re-request review from reviewers when ready

guggero commented on chanbackup/backup.go:
```diff
@@ -53,11 +53,60 @@ func assembleChanBackup(addrSource AddressSource,
+
+// BackupConfig contains options of backup creation process.
+type BackupConfig struct {
+	// Whether to put CloseTxInputs into a backup. A backup with this data
```
Same further below in this commit.

starius replied:
Fixed

guggero reviewed (commented):
Very nice PR, love the commit structure.

Mostly have some nits and concerns around how this is all presented to the user.
But I think the technical implementation makes sense.

Will do a manual round of testing once the review comments from this round have been addressed.

lightninglabs-deploy commented:
@starius, remember to re-request review from reviewers when ready

starius commented on lnwallet/channel.go:
```diff
@@ -6473,29 +6473,68 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
-	localCommit := lc.channelState.LocalCommitment
-	commitTx := localCommit.CommitTx.Copy()
+// SignedCommitTxInputs contains data needed to create a signed commit
+// transaction using signer. See GetSignedCommitTx.
```
Fixed

starius commented on lnwallet/channel.go:
```diff
@@ -6473,29 +6473,68 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
 
-	ourKey := lc.channelState.LocalChanCfg.MultiSigKey
-	theirKey := lc.channelState.RemoteChanCfg.MultiSigKey
+type TaprootSignedCommitTxInputs struct {
```
Fixed

starius commented on lnwallet/channel.go:
```diff
@@ -6473,29 +6473,68 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
+	// the commitment transaction that spends the multi-sig output.
+	SignDesc *input.SignDescriptor
+
+	// If the channel is of taproot type, this is filled.
```
Fixed

starius commented on lnwallet/channel.go:
```diff
@@ -6584,6 +6623,31 @@ func (lc *LightningChannel) getSignedCommitTx() (*wire.MsgTx, error) {
 	return commitTx, nil
 }
 
+// getSignedCommitTx function takes the latest commitment transaction and
```
Fixed

starius commented on chanbackup/backup.go:
```diff
@@ -53,11 +53,60 @@ func assembleChanBackup(addrSource AddressSource,
 	return &single, nil
 }
 
+func buildCloseTxInputs(targetChan *channeldb.OpenChannel) *CloseTxInputs {
```
Fixed

starius commented on chanbackup/backup.go:
```diff
@@ -53,11 +53,60 @@ func assembleChanBackup(addrSource AddressSource,
+
+// WithCloseTxInputs specifies if SCB must contain inputs needed to produce
+// a force close transaction from it using "chantools scbforceclose".
+func WithCloseTxInputs(include bool) BackupOption {
```
Fixed

starius commented on chanbackup/single.go:
```diff
@@ -320,6 +355,33 @@ func (s *Single) Serialize(w io.Writer) error {
 		}
 	}
 
+	version := s.Version
```
I added method `.Encode()` and function `DecodeVersion`. Types `SingleBackupVersion` and `byte` are separate: `SingleBackupVersion` is a clean enum member (without `closeTxVersionMask`) and `byte` is an OR of `SingleBackupVersion` and `closeTxVersionMask` bit. So `s.Version` doesn't have a flag and `s.Version == SimpleTaprootVersion` is a correct comparison. I added a unit tests for the encoding functions.

starius commented on chanbackup/single_test.go:
```diff
@@ -95,6 +96,40 @@ func assertSingleEqual(t *testing.T, a, b Single) {
 	}
+
+	if (a.CloseTxInputs != nil) != (b.CloseTxInputs != nil) {
+		t.Fatalf("closeTxInputs mismatch: %v vs %v",
```
Fixed

starius commented on cmd/lncli/commands.go:
```diff
@@ -2492,7 +2492,8 @@ var exportChanBackupCommand = cli.Command{
 	Usage: "Obtain a static channel back up for a selected channels, " +
 		"or all known channels.",
-	ArgsUsage: "[chan_point] [--all] [--output_file]",
+	ArgsUsage: "[chan_point] [--all] [--with_close_tx_inputs] "+
```
Fixed! I removed RPC options and the LND option. CloseTx is always included into backups.

starius commented on config.go:
```diff
@@ -351,6 +351,8 @@ type Config struct {
 	MaxPendingChannels int    `long:"maxpendingchannels" description:"The maximum number of incoming pending channels permitted per peer."`
 	BackupFilePath     string `long:"backupfilepath" description:"The target location of the channel backup file"`
 
+	BackupCloseTxInputs bool `long:"backupclosetxinputs" description:"Causes the backup to include inputs sufficient to produce the latest force close transaction given private key is available. Causes one additional update of the file, at shutdown time."`
```
I removed the option and made `channelbackuponshutdown` always on.

Channel backup is always updated upon shutdown now. I think this is a good functionality: if some user ever needs to use close tx from channel.backup, it is better to have the latest versions of channels, so it is better to enable this. If we make `channelbackuponshutdown` an option, a regular user is unlikely to enable it and there is a higher probability of channel states in channel.backup to be outdated.

starius commented on server.go:
```diff
@@ -1507,12 +1507,14 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 	backupFile := chanbackup.NewMultiFile(cfg.BackupFilePath)
 	startingChans, err := chanbackup.FetchStaticChanBackups(
 		s.chanStateDB, s.addrSource,
+		chanbackup.WithCloseTxInputs(cfg.BackupCloseTxInputs),
```
Done

starius commented on server.go:
```diff
@@ -4691,6 +4699,47 @@ func (s *server) applyChannelUpdate(update *lnwire.ChannelUpdate,
 
+// updateChannelBackup reads channels from DB, creates channel backup for all
+// the channels and writes it to disk.
+func (s *server) updateChannelBackup() error {
```
`SubSwapper.updateBackupFile()` writes its in-memory state to disk, but its in-memory state is not updated on every channel update, so such a call would be insufficient to write the latest channel states to disk.

I added method `ManualUpdate()` pushing channels to SubSwapper and waiting for it to write them to disk. I call the method from the shutdown code.

starius commented on docs/recovery.md:
```diff
@@ -258,6 +258,25 @@ An example of using file system level notification to [copy the backup to a
 here](https://gist.github.com/alexbosworth/2c5e185aedbdac45a03655b709e255a3).
 
+(backupclosetxinputs)=
+##### backupclosetxinputs
```
Done. Reworked this section.

starius commented on docs/recovery.md:
```diff
@@ -258,6 +258,25 @@ An example of using file system level notification to [copy the backup to a
+perilous. If the channel has been updated since the backup creation, another
+node or a watchtower may issue a penalty transaction, seizing all funds.
+
+Reserve this option as a last resort when the peer is offline and all other
```
Done :+1:

starius commented on docs/release-notes/release-notes-0.18.1.md:
```diff
@@ -39,6 +39,12 @@
   send payment stream context, or automatically at the end of the timeout period 
   if the user provided `timeout_seconds`.
 
+* [`ExportChannelBackup`, `ExportAllChannelBackups`, and
```
Done

starius commented:
@guggero Thanks for the review! 💯

I fixed the comments.

Also I moved function `SignCloseTx` from chantools PR here and used it in itest which tests the following cases:

  - Make sure `channel.backup` file is updated upon LND shutdown.
  - Make sure `channel.backup` file has sufficient information combined with seed and password to produce a force close tx and sweep the funds.
  - Make sure that if the latest channel version is used in `channel.backup` file in scbforceclose recovery method, another peer does not penalize.
  - Control for the previous test case: make sure that if an outdated channel state in `channel.backup` is used, another node can penalize.

These changes are done in two latest commits.

starius commented:
Rebased. Moved the release notes change to 0.18.3.

guggero commented on lnwallet/channel.go:
```diff
@@ -6584,6 +6627,31 @@ func (lc *LightningChannel) getSignedCommitTx() (*wire.MsgTx, error) {
+	localCommit := lc.channelState.LocalCommitment
+
+	inputs := SignedCommitTxInputs{
+		CommitTx:  localCommit.CommitTx,
```
Should use `.CommitTx.Copy()` to retain old behavior.

starius replied:
Is it really needed? `inputs` is passed to function `GetSignedCommitTx` that does `commitTx := inputs.CommitTx.Copy()` and does not modify `inputs.CommitTx`.

starius replied:
I think that functions which modify their arguments (passed through a pointer) should be discouraged. However it is ok for methods like "Unmarshal".

guggero commented on chanbackup/backup.go:
```diff
@@ -53,11 +53,68 @@ func assembleChanBackup(addrSource AddressSource,
+// an open channel. Anyone having these inputs and the signer, can sign the
+// force closure transaction. Warning! If the channel state updates, an attempt
+// to close the channel using this method with outdated CloseTxInputs can result
+// in funds loss!
```
nit: loss of funds.

starius replied:
Fixed

guggero commented on chanbackup/pubsub.go:
```diff
@@ -97,14 +97,25 @@ type SubSwapper struct {
 	quit chan struct{}
 	wg   sync.WaitGroup
+
+	// includeCloseTxInputs specifies if to put CloseTxInputs into a backup.
```
nit: "specifies if we should put the CloseTxInputs into..."?

starius replied:
Fixed. Also, fixed another instance in `chanbackup/backup.go`.

guggero commented on chanbackup/pubsub.go:
```diff
@@ -180,6 +196,44 @@ func (s *SubSwapper) Stop() error {
+	// Wait for processing, block on errChan.
+	var err error
+	select {
+	case err = <-errChan:
```
Handle the error directly in the case statement?

starius replied:
Done :+1:

guggero commented on chanbackup/pubsub.go:
```diff
@@ -321,6 +375,38 @@ func (s *SubSwapper) backupUpdater() {
+			// Send the outcome back to the caller of ManualUpdate.
+			// The error channel must have capacity of 1 not to
+			// block here.
+			manualUpdate.errChan <- err
```
I think we should send the `err` back in the `if err != nil }` above and then just do `close(manualUpdate.errChan)` to signify success.

starius replied:
Done.

guggero commented on chanbackup/pubsub.go:
```diff
@@ -180,6 +196,44 @@ func (s *SubSwapper) Stop() error {
+// ManualUpdate inserts/updates channel states into the swapper. The updates
+// are processed in another goroutine. If wait=true, the method waits for the
+// updates to be fully processed and the file updated on-disk before returning.
+func (s *SubSwapper) ManualUpdate(singles []Single, wait bool) error {
```
I think `wait` is always true in how we currently use this. Should we just remove that param then to simplify things a bit?

starius replied:
Done.

guggero commented on server.go:
```diff
@@ -2307,6 +2308,25 @@ func (s *server) Stop() error {
+			err := s.chanSubSwapper.ManualUpdate(singles, wait)
+			if err != nil {
+				srvrLog.Warnf("Manual update of channel "+
+					"backup failed: %v.", err)
```
micro-nit: remove full stop after `%v`?

starius replied:
Fixed

guggero commented on chanbackup/recover.go:
```diff
@@ -121,3 +129,122 @@ func UnpackAndRecoverMulti(packedMulti PackedMulti,
+	// Each of the keys in our local channel config only have their
+	// locators populate, so we'll re-derive the raw key now.
+	var err error
+	s.LocalChanCfg.MultiSigKey, err = keyRing.DeriveKey(
```
Should we place this into a new variable to avoid mutating the backup given to the function?

starius replied:
The backup (type `Single`) is passed by value, so this line mutates a local copy of the variable.

To improve readability, I changed this line to save the key to a variable and pass it to functions `createSignDesc` and `createTaprootNonceProducer`. So it is clear that the backup variable does not change in the process.

guggero commented on chanbackup/recover.go:
```diff
@@ -121,3 +129,122 @@ func UnpackAndRecoverMulti(packedMulti PackedMulti,
+		SignDesc:  signDesc,
+	}
+
+	if s.Version == SimpleTaprootVersion {
```
Same here re adding a helper method `s.version.IsTaproot()` to also catch future overlay channel types. Here and in `createSignDescs`.

starius replied:
Done.

guggero commented on itest/lnd_channel_backup_test.go:
```diff
@@ -600,6 +660,57 @@ func testChannelBackupRestoreCommitTypes(ht *lntest.HarnessTest) {
+
+		// Test scbforceclose scenario: generate force close tx
+		// from SCB, Carol is offline forever.
+		success = ht.Run("scbforceclose_no_carol/"+tc.name,
```
nit: formatting of multi-argument function call, here and the other cases below.

starius replied:
Fixed.

guggero commented on docs/recovery.md:
```diff
@@ -258,6 +258,22 @@ An example of using file system level notification to [copy the backup to a
+To generate a force close transaction from the backup file, utilize the
+`chantools scbforceclose` command. However, exercise caution as this action is
+perilous. If the channel has been updated since the backup creation, another
+node or a watchtower may issue a penalty transaction, seizing all funds!
```
:+1:

guggero reviewed (commented):
Amazing work on this! Code looks really good now :100: 

Still want to do some manual tests, which I didn't get to yet. But will fully approve afterward. So I think it's a good time for a second reviewer to look at this now.

guggero reviewed (approved):
tACK, LGTM :tada:

starius commented on chanbackup/single.go:
```diff
@@ -320,6 +381,33 @@ func (s *Single) Serialize(w io.Writer) error {
+			return err
+		}
+
+		if s.Version == SimpleTaprootVersion {
```
Done :100:

starius commented on itest/hdsigner.go:
```diff
@@ -0,0 +1,197 @@
+	// Chop off the sighash flag at the end of the signature.
+	return ecdsa.ParseDERSignature(sig[:len(sig)-1])
+}
+func (s *hdSigner) ComputeInputScript(_ *wire.MsgTx, _ *input.SignDescriptor) (
```
Fixed.

starius commented on itest/hdsigner.go:
```diff
@@ -0,0 +1,197 @@
+			"key: %w", err)
+	}
+
+	return ecdh(privKey, pubKey)
```
Done.

starius commented:
@guggero Thank you for review! I addresses the feedback.

Also, I checked the failures in CI. The failures of itests on windows and "unit tests (btcd unit-cover)" seem to be unrelated to the PR.

One of new itests [failed](https://github.com/lightningnetwork/lnd/actions/runs/9921314522/job/27503455107?pr=8183) on neutrino backend:

```
    lnd_channel_backup_test.go:290: 
        	Error Trace:	/home/runner/work/lnd/lnd/itest/lnd_channel_backup_test.go:2287
        	            				/home/runner/work/lnd/lnd/itest/lnd_channel_backup_test.go:290
        	            				/home/runner/work/lnd/lnd/itest/lnd_channel_backup_test.go:940
        	            				/home/runner/work/lnd/lnd/itest/lnd_channel_backup_test.go:705
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 0
        	Test:       	TestLightningNetworkDaemon/tranche00/07-of-156/neutrino/channel_backup_restore_commit_types/scbforceclose_carol_penalizes/restore_from_backup_file_anchors
```

This itests is passing on my machine. It could happen in CI, because Carol didn't manage to broadcast the penalty transaction in time before `ht.MineBlocksAndAssertNumTxes(1, 1)` was called, so the mined block contained no transactions. Race condition. Might this be the case?

I added `require.Eventually` waiting for mempool for having 1 transaction before `MineBlocksAndAssertNumTxes`.

There are other `MineBlocksAndAssertNumTxes` in itests. I'm curious if they might be race conditions?

guggero commented:
> Race condition. Might this be the case?

Yes, unfortunately the neutrino itest isn't super stable on CI yet, mainly due to force close timing issues. I think @yyforyongyu is looking into that with https://github.com/lightningnetwork/lnd/pull/8892.

So if the test passes for you locally, that's fine.

starius commented:
Neutrino itests are passing now in CI.

Windows itests fail, but they are flacky in master branch. Failures seem to be unrelated to the PR.

btcd unit-cover tests fail in master with the same error:
```
Error: cannot read source of file "/home/runner/work/lnd/lnd/fn/func.go": open /home/runner/work/lnd/lnd/fn/func.go: no such file or directory
```

yyforyongyu commented on itest/hdkeychain.go:
```diff
@@ -0,0 +1,88 @@
+	return currentKey, nil
+}
+
+type hdKeyRing struct {
```
why do we need this `hdKeyRing`? By searching `DeriveNextKey` it seems we've already had several implementations already, can we use one of those instead? Also isn't there an RPC service `RPCKeyRing` we can use?

starius replied:
Can you elaborate on this, please? I can't find a suitable in-memory implementation of keychain.KeyRing, keychain.ECDHRing, and input.Signer (the types I need in SignCloseTx) which could be created from seed and passphrase.

To create RPCKeyRing, the following objects are needed: keychain.SecretKeyRing, lnwallet.WalletController, and *lncfg.RemoteSigner. I don't know how to make them from seed and passphrase.

I copy-pasted these two types from chantools. If there is a better way to create them, we can consider using that approach in chantools as well.

yyforyongyu commented on itest/hdsigner.go:
```diff
@@ -0,0 +1,176 @@
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+type hdSigner struct {
```
same question re this, why do we need another `input.Signer`?

starius replied:
Same as above: I don't know how to create `input.Signer` from seed and passphrase using existing types :)

yyforyongyu commented on itest/lnd_channel_backup_test.go:
```diff
@@ -665,6 +782,170 @@ func runChanRestoreScenarioCommitTypes(ht *lntest.HarnessTest,
+		// Use the latest SCB to produce force close tx.
+		scbForCloseTx = backupAfterShutdown
+	}
+	crs.forceCloseTx, err = chanbackup.SignCloseTx(
```
Instead of creating key rings and signers outside of the `SignCloseTx`, I think it should be handled inside the method. This way we only need to pass `scbForCloseTx` and `mnemonic` and `password`, and the `chanbackup` handles the creation of signers and king rings.

Also by looking at this method alone it's not an RPC point? If it's a method meant to be imported by other packages to be used, I'm not sure if it should be tested in integration tests.

starius replied:
If we make it too easy for the user to make signed transactions, they are more likely to misuse the feature and get funds lost. Currently it is not trivial to use the function outside of chantools, where we warn the user many times. The implementations of dependencies (keychain.KeyRing, keychain.ECDHRing, and input.Signer) live as private types in itest and are not directly usable from outside.

This function is intended to be used from chantools only. We don't have chantools in itest, so the only way of emulating using chantools in itest is to call the function directly. The code in chantools is really a wrapper around this function: https://github.com/lightninglabs/chantools/pull/95/commits/1fe8248114f0c6f453ab28b99b3d74abdcf1943b

yyforyongyu replied:
In that case we should add unit tests for this method, not itest. You can import the `lntest` package in `chantools` to build the exact same test, and we should not use the itest in `lnd` to test programs importing it.

starius replied:
@yyforyongyu I removed most itests (using hdsigner and hdkeyring). They were replaced with unit tests in chantools. Now the only itest change left is the check that LND updates channel.backup file upon shutdown, which does not add any new dependencies.

yyforyongyu commented on itest/lnd_channel_backup_test.go:
```diff
@@ -1673,3 +1954,389 @@ func assertDLPExecuted(ht *lntest.HarnessTest,
+	ht.EnsureConnected(carol, dave)
+
+	// Wait for Carol to publish a penalty transaction.
+	require.Eventually(ht, func() bool {
```
I think `MineBlocksAndAssertNumTxes` already waits the tx?

starius replied:
Great catch!

I added `require.Eventually` because I got a failure in CI in `ht.MineBlocksAndAssertNumTxes(1, 1)` call (it got 0 txs), I [wrote a comment about it](https://github.com/lightningnetwork/lnd/pull/8183#issuecomment-2232295216) and thought mistakenly that `MineBlocksAndAssertNumTxes` doesn't wait.

I checked the code of `MineBlocksAndAssertNumTxes` and `AssertNumTxsInMempool` and now I see that it was not the reason of that failure, because `AssertNumTxsInMempool` waits for 1 minute. I don't know why there were no transactions in mempool in that run of the test. Penalty transaction has to be there... I'll try to undo `require.Eventually` and reproduce it locally and debug.

starius replied:
I rebased the PR and fixed itest build errors (changes in ht interface and chanRestoreViaRPC function).

Also removed waiting for tx in mempool and added waiting for ClosedChannels call next to it, since it failed sometimes, because a channel was still closing during the call (so it saw 0 closed channels).

However I can't reproduce the original failure locally. For some reason Carol node didn't issue a justice transaction in that run in CI. Do you have any ideas why it could happen?

yyforyongyu reviewed (changes requested):
Nice tests! I only looked at the itest related changes and have several questions.

starius commented:
I rebased the PR and fixed itest build errors (changes in ht interface and chanRestoreViaRPC function).

Also removed waiting for tx in mempool and added waiting for ClosedChannels call next to it, since it failed sometimes, because a channel was still closing during the call (so it saw 0 closed channels).

Added commit "lntest: fix typo".

lightninglabs-deploy commented:
@bhandras: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bhandras: review reminder
@starius, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@starius, remember to re-request review from reviewers when ready

ziggie1984 commented on itest/lnd_channel_backup_test.go:
```diff
@@ -214,24 +228,71 @@ func (c *chanRestoreScenario) testScenario(ht *lntest.HarnessTest,
-	// let's start up Carol again.
-	require.NoError(ht, restartCarol(), "restart carol failed")
+	if c.forceCloseTx == nil {
+		// DLP case.
```
Nit: Move the comment above the if clause

starius replied:
Fixed

ziggie1984 commented on itest/lnd_channel_backup_test.go:
```diff
@@ -1672,3 +1954,383 @@ func assertDLPExecuted(ht *lntest.HarnessTest,
+
+	// Upon reconnection, the nodes should detect that Dave is out of sync.
+	// Carol should force close the channel using her latest commitment.
+	// For scbforceclose case: we detect the manually broadcasts Dave's tx.
```
According to the test it looks to me that  `assertFundsRecoveredStartCarol ` is only called in the scbforceclose case ?

starius replied:
Fixed. It was a relict of my original approach trying to have a single "assert" function for all cases. The function became too branched and complicated and I split it into separate "assert" functions, but forgot to update this comment.

ziggie1984 commented on itest/lnd_channel_backup_test.go:
```diff
@@ -1672,3 +1954,383 @@ func assertDLPExecuted(ht *lntest.HarnessTest,
+		// are uneconomical.
+		ht.AssertNumPendingSweeps(dave, 1)
+
+		// Carol has two sweeps.
```
can you add which sweep Carol has here in addition to the anchor ?

starius replied:
Done. 

```
// Carol has two sweeps: an anchor and her balance from the
// channel. Carol's funds are available immediately, as they are
// not timelocked.
```

ziggie1984 commented on itest/lnd_channel_backup_test.go:
```diff
@@ -1672,3 +1954,383 @@ func assertDLPExecuted(ht *lntest.HarnessTest,
+
+		// After Carol's output matures, she should also reclaim her
+		// funds.
+		//
+		// The commit sweep resolver publishes the sweep tx at
+		// defaultCSV-1 and we already mined one block after the
+		// commitmment was published, so take that into account.
+		ht.MineEmptyBlocks(int(defaultCSV - blocksMined))
+
+		// Carol should have two sweep requests - one for her commit
+		// output and the other for her anchor.
+		ht.AssertNumPendingSweeps(carol, 2)
+
+		// Mine a block to trigger the sweep in Carol.
+		ht.MineEmptyBlocks(1)
+		ht.MineBlocksAndAssertNumTxes(1, 1)
+
+		// Now the channel should be fully closed also from
+		// Carol's POV.
+		ht.AssertNumPendingForceClose(carol, 0)
+
+		// We'll now mine the remaining blocks to prompt Dave to sweep
+		// his CLTV-constrained output.
+		resp := dave.RPC.PendingChannels()
+		blocksTilMaturity :=
+			resp.PendingForceClosingChannels[0].BlocksTilMaturity
+		require.Positive(ht, blocksTilMaturity)
+
+		ht.MineEmptyBlocks(int(blocksTilMaturity))
+
+		// Dave should have two sweep requests - one for his commit
+		// output and the other for his anchor.
+		ht.AssertNumPendingSweeps(dave, 2)
+
+		// Mine a block to trigger the sweep.
+		ht.MineEmptyBlocks(1)
+		ht.MineBlocksAndAssertNumTxes(1, 1)
+
+		// Now Dave should consider the channel fully closed.
+		ht.AssertNumPendingForceClose(dave, 0)
```
I think there is no need to test the sweeper here again imo, checking whether the sweeps where requested is sufficient enough.

starius replied:
Fixed. Removed extra checks.

ziggie1984 commented on itest/lnd_channel_backup_test.go:
```diff
@@ -1672,3 +1954,383 @@ func assertDLPExecuted(ht *lntest.HarnessTest,
+		// Mine one block to trigger the sweeper to sweep.
+		ht.MineEmptyBlocks(1)
+		blocksMined++
+
+		// Expect one tx - the commitment sweep from Carol. For
+		// anchor channels, we expect the two anchor sweeping
+		// txns to be failed due they are uneconomical.
+		ht.MineBlocksAndAssertNumTxes(1, 1)
+		blocksMined++
+
+		// Now Carol should consider the channel fully closed.
+		ht.AssertNumPendingForceClose(carol, 0)
+
+		// After Dave's output matures, he should also reclaim his
+		// funds.
+		//
+		// The commit sweep resolver publishes the sweep tx at
+		// defaultCSV-1 and we already have blocks mined after the
+		// commitmment was published, so take that into account.
+		ht.MineEmptyBlocks(int(defaultCSV - blocksMined))
+
+		// Mine one block to trigger the sweeper to sweep.
+		ht.MineEmptyBlocks(1)
+
+		// Dave's sweeper swept the outputs of force close tx.
+		if lntest.CommitTypeHasAnchors(commitType) {
+			ht.AssertNumPendingSweeps(dave, 2)
+		} else {
+			ht.AssertNumPendingSweeps(dave, 1)
+		}
+
+		// Assert the sweeping tx is mined.
+		ht.MineBlocksAndAssertNumTxes(1, 1)
+
+		// Funds are now in Dave's wallet.
+		ht.AssertNumPendingForceClose(dave, 0)
+
+		// Now the channel should be fully closed also from
+		// Carol's POV.
+		ht.AssertNumPendingForceClose(carol, 0)
```
Same here, I think we can reduce the test quite a bit, because if something changes for example with the anchor sweep all those tests would need to be changed, so I think we should not repeat tests wdyt ?

starius replied:
Fixed

ziggie1984 reviewed (commented):
Great work on this feature, definitely  important work 👌

Well thought through code design only had minor comment on this one.

Regarding the itest however I am in doubt whether we need this test in general because it introduces a lot of new code regarding the signer, keyring etc. Moreover I think most of the tests here are already covered in other itests, for example the sweeping logic or the penalty case. Would it maybe possible to move all this tests into the chantools repo instead ? 
I want to outline that these itests definiitly took a lot of work and are very detailed in general so kudos for this work 💪. I am brining this discussion up because when for example changing the anchor sweeper logic we realised that we had to change the itests in so many different places which took a lot of work and IMO was quite unnecessary. Thats why I want to avoid to many itests which basically test the same thing.

ziggie1984 commented:
Would love to see a follow-up PR with the HTLC-Sigs included for the outgoing HTLCs 😊

yyforyongyu reviewed (changes requested):
If we decide to support the feature of signing close tx in `lnd`, then we should make it into a RPC endpoint. If we decide to only make it a method that can be imported by other projects, i.e., the `SignCloseTx` method introduced here, then it should be only affecting the module `chanbackup` (ideally we should package it but that's beyond the scope), which means we should write unit tests, not itest.

We can easily duplicate the itest written in this PR since `lntest` is made [for that](https://github.com/lightningnetwork/lnd/tree/master/lntest#quick-start). This is also how other services like loop/naut build their tests, not the other way around.

lightninglabs-deploy commented:
@starius, remember to re-request review from reviewers when ready

starius commented on lnwallet/channel.go:
```diff
@@ -6473,29 +6473,72 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
+	SignDesc *input.SignDescriptor
+
+	// Taproot holds fields needed in case of a taproot channel.
+	// If and only if the channel is of taproot type, this field is filled.
```
Done

starius commented on lnwallet/channel.go:
```diff
@@ -6473,29 +6473,72 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
+	CommitSig []byte
+
+	// OurKey is our key to be used within the 2-of-2 output script
+	// for the owner of this channel config.
```
Yes, this was a copy-paste relict. Fixed.

starius commented on lnwallet/channel.go:
```diff
@@ -6473,29 +6473,72 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
+	// the total number of commitment updates to this point. This can be
+	// viewed as sort of a "commitment height" as this number is
+	// monotonically increasing.
+	CommitHeight uint64
```
Yes. I added a comment in the code:

```go
        // CommitHeight is the update number that this ChannelDelta represents
        // the total number of commitment updates to this point. This can be
        // viewed as sort of a "commitment height" as this number is
        // monotonically increasing. This number is used to make a signature
        // for a taproot channel, since it is used by shachain nonce producer
        // (TaprootNonceProducer).
        CommitHeight uint64
```

starius commented on lnwallet/channel.go:
```diff
@@ -6473,29 +6473,72 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
 	return lc.channelState.AbsoluteThawHeight()
 }
 
-// getSignedCommitTx function take the latest commitment transaction and
```
We store their part of the signature in the field `CommitSig`.

starius commented on chanbackup/single_test.go:
```diff
@@ -195,6 +217,45 @@ func genRandomOpenChannelShell() (*channeldb.OpenChannel, error) {
 	}, nil
 }
 
+// TestVersionEncoding tests encodign and decoding of version byte.
```
Fixed

starius commented on chanbackup/pubsub.go:
```diff
@@ -180,6 +196,39 @@ func (s *SubSwapper) Stop() error {
 
+// ManualUpdate inserts/updates channel states into the swapper. The updates
+// are processed in another goroutine. The method waits for the updates to be
+// fully processed and the file updated on-disk before returning.
```
Fixed

starius commented on chanbackup/pubsub.go:
```diff
@@ -321,6 +370,42 @@ func (s *SubSwapper) backupUpdater() {
+				"num_old_chans=%v, num_new_chans=%v",
+				oldStateSize, newStateSize)
+
+			// With out new state constructed, we'll, atomically
```
Fixed

starius commented on chanbackup/pubsub.go:
```diff
@@ -321,6 +370,42 @@ func (s *SubSwapper) backupUpdater() {
+					err)
+
+				// Send the error to the caller of ManualUpdate.
+				// The error channel must have capacity of 1 not
```
Do you mean to make sure that the channel does have capacity of 1? We can send to the channel in select and log an error or `panic` in the default case. What do you think?

ziggie1984 replied:
Could you explain why we cannot block here, backupUpdater runs in its own goroutine, why not sync this call via the errChan with the server func?

starius replied:
`ManualUpdate` receives from `errChan` and `s,quit` in `select`. If `s.quit` is closed while waiting, then nobody reads from `errChan` and `backupUpdater` goroutine leaks. This can be worked around by using another `select` to send to `errChan` and receive from `s.quit` at the same time, but I like the approach with a buffered channel more, because it is harder to make a dead-lock this way.

ziggie1984 replied:
thanks for the explantion: 

just do `manualUpdate.errChan <- s.updateBackupFile()` here.

Moreover add a comment why we need to be buffered when creating the errChan in `ManualUpdate`

starius commented on itest/lnd_channel_backup_test.go:
```diff
@@ -600,6 +661,63 @@ func testChannelBackupRestoreCommitTypes(ht *lntest.HarnessTest) {
+		}
+
+		// Control for the previous case: make sure that Carol can
+		// produce a penalty transaction is needed. To do it, use an
```
It was a typo "if needed"
Fixed

starius commented on chanbackup/backup.go:
```diff
@@ -53,11 +53,68 @@ func assembleChanBackup(addrSource AddressSource,
 
+// buildCloseTxInputs generates inputs needed to force close a channel from
+// an open channel. Anyone having these inputs and the signer, can sign the
+// force closure transaction. Warning! If the channel state updates, an attempt
```
This may happen if an outdated channel backup is attempted to be used to force close the channel.

I added this note to the comment.

starius commented on chanbackup/backup.go:
```diff
@@ -53,11 +53,68 @@ func assembleChanBackup(addrSource AddressSource,
+		CommitSig: localCommit.CommitSig,
+	}
+
+	// In case of a taproot channel, commit height is needed as well.
```
It is needed to produce verification nonce for the taproot channel using shachain.

I added this note to the comment.

starius commented on chanbackup/pubsub.go:
```diff
@@ -97,14 +97,25 @@ type SubSwapper struct {
 	quit chan struct{}
 	wg   sync.WaitGroup
+
+	// includeCloseTxInputs specifies if we should put the CloseTxInputs
```
This was a relict of previous PR version in which we had this functionality under an option. We decided to have it always on, but this place remained with options. I reworked it, so it is now also always on.

starius commented on chanbackup/single.go:
```diff
@@ -52,8 +52,44 @@ const (
 )
 
+// Encode returns encoding of the version to put into channel backup.
+// Argument closeTx specifies of the backup includes force close transaction.
```
Fixed

starius commented on chanbackup/single.go:
```diff
@@ -52,8 +52,44 @@ const (
+}
+
+// Decode decodes the encoding of the version from a channel backup.
+// It returns the version and if the backup includes force close transaction.
```
Fixed

starius commented on chanbackup/single.go:
```diff
@@ -52,8 +52,44 @@ const (
+	closeTx := (encoded & closeTxVersionMask) != 0
+
+	// Remove the closeTx bit.
+	version := SingleBackupVersion(encoded &^ closeTxVersionMask)
```
Fixed.

```go
        // The encoded form has version byte as bit OR of SingleBackupVersion
        // value and closeTx bit. So here we remove closeTx bit.
        version := SingleBackupVersion(encoded &^ closeTxVersionMask)
```

starius replied:
Fixed

starius commented on chanbackup/single.go:
```diff
@@ -138,6 +174,37 @@ type Single struct {
+	// The field is optional.
+	//
+	// See WithCloseTxInputs.
+	CloseTxInputs *CloseTxInputs
```
Done.

This PR was opened before fn.Option was added :)

starius commented on chanbackup/single.go:
```diff
@@ -320,6 +387,33 @@ func (s *Single) Serialize(w io.Writer) error {
+	// Encode version enum and hasCloseTx flag to version byte.
+	version := s.Version.Encode(s.CloseTxInputs != nil)
+
+	if s.CloseTxInputs != nil {
```
Done

starius commented on chanbackup/pubsub.go:
```diff
@@ -180,6 +196,39 @@ func (s *SubSwapper) Stop() error {
+// ManualUpdate inserts/updates channel states into the swapper. The updates
+// are processed in another goroutine. The method waits for the updates to be
+// fully processed and the file updated on-disk before returning.
+func (s *SubSwapper) ManualUpdate(singles []Single) error {
```
I want the method to have convenient synchronous interface, because it is used in method `server.Stop`, which itself is quite complicated.

starius commented on server.go:
```diff
@@ -2332,6 +2333,24 @@ func (s *server) Stop() error {
 			srvrLog.Warnf("failed to stop htlcNotifier: %v", err)
 		}
+
+		// Update channel.backup file. Make sure to do it before
```
The primary goal of updating `channel.backup` before shutdown is to update the commitment transaction to the most recent version. Because of this, if LND was shutdown properly, the versions of the close transactions in the backup are always the latest ones and it is safe to use that channel backup to force close a channel if LND hasn't started after that. This is a convenient property for those who runs LND only when needed. They can backup `channel.backup` after each LND using (and shurring down) and be self-soverein in their funds.

ziggie1984 replied:
TBH, I think it also makes sense to add a timestamp here making it more tangible for a user whether this is an outdated closeTx

starius commented on chanbackup/recover.go:
```diff
@@ -121,3 +130,129 @@ func UnpackAndRecoverMulti(packedMulti PackedMulti,
+	}
+
+	// Each of the keys in our local channel config only have their
+	// locators populate, so we'll re-derive the raw key now.
```
Fixed

starius commented on itest/hdkeychain.go:
```diff
@@ -0,0 +1,88 @@
+package itest
```
Fixed. Also in file itest/hdsigner.go

starius commented on itest/hdkeychain.go:
```diff
@@ -0,0 +1,88 @@
+
+	var currentKey = key
+	for idx, pathPart := range path {
+		derivedKey, err := currentKey.DeriveNonStandard(pathPart)
```
Interesting! I copy-pasted this code from [chantools](https://github.com/lightninglabs/chantools/blob/master/lnd/hdkeychain.go). If I replace it with `Derive`, itest passes. IIUC there was a bug affecting small percentage of keys for `DeriveNonStandard` (a.k.a. `Child`).

Should we use `Derive` here and also in `chantools`?

ziggie1984 replied:
I think this is not relevant here anymore.

starius commented on itest/hdkeychain.go:
```diff
@@ -0,0 +1,88 @@
+	return currentKey, nil
+}
+
+type hdKeyRing struct {
```
Fixed. I added compiler interface checks here and in itest/hdsigner.go

starius commented on itest/lnd_channel_backup_test.go:
```diff
@@ -161,6 +170,11 @@ func (c *chanRestoreScenario) restoreDave(ht *lntest.HarnessTest,
 //  6. assert DLP is executed.
+//
+// If forceCloseTx is provided, Carol is not restarted and that transaction is
+// broadcasted instead. This is needed to test scbforceclose.
```
This is yet to be released command of chantools which will produce force close transaction from channel.backup using the code from this LND PR. The PR to chantools is here: https://github.com/lightninglabs/chantools/pull/95

The command is mentioned in lnd/docs/recovery.md (I'm adding in this PR in commit "docs/recovery: add Last resort manual force close").

starius commented:
I rebased the PR, made `chanbackup` package TapscriptRoot channels aware. Added new backup version for TapscriptRoot channels and made changes needed to force-close such channels.

I addressed @ziggie1984 's feedback, in particular used fn.Option for optional parts of the backup and removed includeCloseTxInputs from `chanbackup` code. It is always on for new backups generated.

starius commented:
@yyforyongyu I removed function SignCloseTx and the itest from this PR.
I added added the function SignCloseTx to chantools and added unit tests there: https://github.com/lightninglabs/chantools/pull/95

Please take another look!

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@ziggie1984: review reminder
@starius, remember to re-request review from reviewers when ready

ziggie1984 reviewed (changes requested):
Thank you for incorporating my comments need to also look into the chantools PRs.

Tested this with a litd version and encountered a bug which needs to be fixed. Maybe a unitest encoding a chanbackup and encoding it is enough, or even an itest (but not sure)

Moreover I think we should include a timestamp into the backup now, because that would give users more orientation whether this backup is outdated.

In addition chantools command reading the new version of the backup needs to be updated.

Keep up the great work !

starius commented on chanbackup/pubsub.go:
```diff
@@ -168,6 +184,39 @@ func (s *SubSwapper) Stop() error {
+		errChan: errChan,
+	}
+
+	select {
```
In order to do it, we need `singles`. In case of ManualUpdate, they are passed by the caller. If we add time based updates, we need to run the timer outside of `SubSwapper`, because it does not have the source of this information (`server.chanStateDB` and `server.addrSource`).

ziggie1984 replied:
ahh ok not super trivial gottya

starius commented on chanbackup/single.go:
```diff
@@ -56,8 +58,51 @@ const (
+
+	// closeTxVersionMask is the byte mask used that is ORed to version byte
+	// on wire indicating that the backup has CloseTxInputs.
+	closeTxVersionMask = 1 << 7
```
I'll do it after adding the timestamp field in this PR.

starius commented on chanbackup/single.go:
```diff
@@ -142,11 +187,47 @@ type Single struct {
+
+	// TapscriptRoot is the root of the tapscript tree that will be used to
+	// create the funding output. This is an optional field that should
+	// only be set for taproot channels (HasTapscriptRoot).
```
Fixed

starius commented on chanbackup/single.go:
```diff
@@ -142,11 +187,47 @@ type Single struct {
 	LeaseExpiry uint32
+
+	// CloseTxInputs contains data needed to produce a force close tx
+	// using "chantools scbforceclose".
```
Fixed

starius commented on chanbackup/backup.go:
```diff
@@ -53,6 +54,43 @@ func assembleChanBackup(addrSource AddressSource,
+
+	// In case of a taproot channel, commit height is needed as well to
+	// produce verification nonce for the taproot channel using shachain.
+	if targetChan.ChanType.IsTaproot() {
```
Great catch! I added an `if` to pass `TapscriptRoot` and modified the test (genRandomOpenChannelShell) to generate data exposing this bug.

I prefer `if` over `switch` here, because the cases overlap: for a custom channel `IsTaproot()` is also true, so the proposed code actually is not working: `switch` jumps into the first branch for a custom channel as well as for simple taproot channel.

starius commented on chanbackup/single.go:
```diff
@@ -52,6 +52,10 @@ const (
 	// SimpleTaprootVersion is a version that denotes this channel is using
 	// the musig2 based taproot commitment format.
 	SimpleTaprootVersion = 5
+
```
Done. Commit msg is now as follows:

```
    chanbackup: add backup version for TapscriptRoot
    
    Previous to this change taproot assets channels and simple taproot channels were
    considered the same in the context of chanbackup package, since they stored the
    same data. In the following commits we are adding the data needed to produce a
    signed commitment transaction from a SCB file and in order to do that we need to
    add more fields and a custom channel gets one additional field (TapscriptRoot)
    compared to a simple taproot channel. So now we have to distinguish these kinds
    of channels in chanbackup package.
    
    See PR https://github.com/lightningnetwork/lnd/pull/8183 for more details.
```

starius commented on lnwallet/channel.go:
```diff
@@ -6342,49 +6342,100 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
+// GetSignedCommitTx function takes the latest commitment transaction and
+// populates it with witness data. This pure function is needed to be able
+// to store the inputs without a signature in SCB and sign the transaction
+// in chantools scbforceclose.
```
Applied

starius commented on lnwallet/channel.go:
```diff
@@ -6342,49 +6342,100 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
+// See GetSignedCommitTx.
+type TaprootSignedCommitTxInputs struct {
+	// CommitHeight is the update number that this ChannelDelta represents
+	// the total number of commitment updates to this point. This can be
```
Applied

starius commented on lnwallet/channel.go:
```diff
@@ -6342,49 +6342,100 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
+
+	// Taproot holds fields needed in case of a taproot channel.
+	// Iff the channel is of taproot type, this field is filled.
+	Taproot *TaprootSignedCommitTxInputs
```
Done

Roasbeef commented on chanbackup/single.go:
```diff
@@ -56,8 +58,51 @@ const (
+}
+
+// IsTaproot returns if this is a backup of a taproot channel. This will also be
+// true for simple taproot overlay channels when a version is added.
```
So we'd handle the SCB process of those with distinct flows. For the normal taproot channel type, the sweeping process is more or less the same. For the overlay channel type, the SCB needs additional information to be able to sweep the funds, which is out of the scope of this PR.

ziggie1984 replied:
ok makes sense. I would say it also depends on what is merged first. Seems like most of the custom channel stuff is already on master, give that I think we should include this change also in this PR ? Maybe I am mistaken but seems like only PR 5/5 is missing for the custom channel topic, so I think it's good to include it in this PR ?

ziggie1984 reviewed (commented):
Looks good - I think before we get this in we still need to:

1. Confirm that custom channels are backuped
2. Make sure chantools are up-to-date (`dumpchanbackup`)
3. Maybe create a followup to introduce the timestamp.

ziggie1984 reviewed (approved):
1. ✅, successfully verified that also custom-channels are backuped correctly, [see testdata](https://github.com/lightninglabs/chantools/pull/95#issuecomment-2387152027) 
2. dumpchanbackup is handled here: https://github.com/lightninglabs/chantools/pull/95 => must be merged when this PR is merged.
3. @starius please open a tracking issue related to the timestamp.

starius commented:
I added small commit "itest: test channel.backup changes upon shutdown" in which I add a check in SCB itest that channel.backup file content changes upon node shutdown. This is needed to make sure that this feature actually works. Just few lines of code in itest, no new dependencies.

starius commented:
@ziggie1984
timestamps issue opened: https://github.com/lightningnetwork/lnd/issues/9149

ziggie1984 commented on itest/lnd_channel_backup_test.go:
```diff
@@ -786,8 +786,15 @@ func runChanRestoreScenarioForceClose(ht *lntest.HarnessTest, zeroConf bool) {
+	// Shutdown to keep its directory including channel.backup file.
+	ht.SuspendNode(dave)
+
+	// Read Dave's channel.backup file again to make sure it was updated
```
Could you add in the comment why it needs to be updated, it will only update if the state changes, otherwise the closeInput info also remains the same ?

starius replied:
Done.

```
        // Read Dave's channel.backup file again to make sure it was updated
        // upon Dave's shutdown. In case LND state is lost and DLP protocol
        // fails, the channel.backup file and the commit tx in it are the
        // measure of last resort to recover funds from the channel. The file
        // is updated upon LND server shutdown to update the commit tx just in
        // case it is used this way. If an outdated commit tx is broadcasted,
        // the funds may be lost in a justice transaction. The file is encrypted
        // and we can't decrypt it here, so we just check that the content of
        // the file has changed.
        multi2, err := os.ReadFile(backupFilePath)
        require.NoError(ht, err)
        require.NotEqual(ht, multi, multi2)
```

> it will only update if the state changes, otherwise the closeInput info also remains the same

The file is encrypted and it changes even if the content is the same.

We can't decrypt it in itest to check closeInputs (this would bring hdkeyring.go dependency which we don't want in this PR).

So I just checked that the file content has changed -- this means ManualUpdate was at least triggered.

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@starius, remember to re-request review from reviewers when ready

yyforyongyu commented on contractcourt/commit_sweep_resolver.go:
```diff
@@ -20,9 +20,15 @@ import (
-// output paying to us, in the case that the remote party broadcasts their
-// version of the commitment transaction. We can sweep this output immediately,
-// as it doesn't have a time-lock delay.
+// output paying to us (local channel balance). In the case that the local
```
👍

yyforyongyu commented on lnwallet/channel.go:
```diff
@@ -6342,49 +6342,105 @@ func (lc *LightningChannel) AbsoluteThawHeight() (uint32, error) {
+	case inputs.Taproot.IsSome():
+		// Extract Taproot from fn.Option. It is safe to call
+		// UnsafeFromSome because we just checked that it is some.
+		taproot := inputs.Taproot.UnsafeFromSome()
```
non-blocking: we prolly need to move this method to a `_dev.go` file and only build it when it's in a test env, otherwise it's very tempting to use this API cc @ProofOfKeags

yyforyongyu commented on chanbackup/backup.go:
```diff
@@ -53,6 +54,48 @@ func assembleChanBackup(addrSource AddressSource,
+// in loss of funds! This may happen if an outdated channel backup is attempted
+// to be used to force close the channel.
+func buildCloseTxInputs(targetChan *channeldb.OpenChannel,
+) fn.Option[CloseTxInputs] {
```
line wrapping needs to be fixed, [reference](https://github.com/lightningnetwork/lnd/blob/master/docs/code_formatting_rules.md#wrapping-long-function-definitions)

starius replied:
Fixed

yyforyongyu commented on chanbackup/single.go:
```diff
@@ -56,8 +58,51 @@ const (
+	return encoded
+}
+
+// Decode decodes the encoding of the version from a channel backup.
```
`Decode` -> `DecodeVersion`

starius replied:
Fixed

yyforyongyu commented on chanbackup/single.go:
```diff
@@ -329,6 +419,56 @@ func (s *Single) Serialize(w io.Writer) error {
+		if err != nil {
+			return
+		}
+
```
nit: could do `!s.Version.IsTaproot()` to save some indentations here

starius replied:
Done. Also added a comment to explain what happens after the check:

```go
                if !s.Version.IsTaproot() {
                        return
                }

                // Write fields needed for taproot channels.
                err = lnwire.WriteElements(
                        &singleBytes, inputs.CommitHeight,
                )
                if err != nil {
                        return
                }
                ...
```

yyforyongyu commented on chanbackup/single.go:
```diff
@@ -543,6 +685,47 @@ func (s *Single) Deserialize(r io.Reader) error {
 		}
 	}
 
+	if hasCloseTx {
```
nit: same here, could do,
```go
if !hasCloseTx {
    return nil
}
```

starius replied:
Done. Also added a comment:

```go
        if !hasCloseTx {
                return nil
        }

        // Deserialize CloseTxInputs if it is present in serialized data.
        commitTx := &wire.MsgTx{}
        if err := commitTx.Deserialize(r); err != nil {
                return err
        }
        ...
```

yyforyongyu reviewed (commented):
Found a few typo/line wrapping issues, otherwise LGTM🙏

starius commented:
@yyforyongyu Thanks for review! I addressed the feedback.
Please take another look!

guggero commented on chanbackup/single.go:
```diff
@@ -329,6 +419,59 @@ func (s *Single) Serialize(w io.Writer) error {
+
+	// Serialize CloseTxInputs if it is provided. Fill err if it fails.
+	var err error
+	s.CloseTxInputs.WhenSome(func(inputs CloseTxInputs) {
```
nit: We could use `fn.MapOptionZ()` here to allow returning an error, which would follow the intended "no mutations within callbacks" paradigm of the `fn` library:

```
	// Serialize CloseTxInputs if it is provided. Fill err if it fails.
	err := fn.MapOptionZ(s.CloseTxInputs, func(inputs CloseTxInputs) error {
		err := inputs.CommitTx.Serialize(&singleBytes)
		if err != nil {
			return err
		}

		err = lnwire.WriteElements(
			&singleBytes,
			uint16(len(inputs.CommitSig)), inputs.CommitSig,
		)
		if err != nil {
			return err
		}

		if !s.Version.IsTaproot() {
			return nil
		}

		// Write fields needed for taproot channels.
		err = lnwire.WriteElements(
			&singleBytes, inputs.CommitHeight,
		)
		if err != nil {
			return err
		}

		if s.Version.HasTapscriptRoot() {
			opt := inputs.TapscriptRoot
			var tapscriptRoot chainhash.Hash
			tapscriptRoot, err = opt.UnwrapOrErr(
				errEmptyTapscriptRoot,
			)
			if err != nil {
				return err
			}

			err = lnwire.WriteElements(
				&singleBytes, tapscriptRoot[:],
			)
			if err != nil {
				return err
			}
		}

		return nil
	})
	if err != nil {
		return fmt.Errorf("failed to encode CloseTxInputs: %w", err)
	}
```

starius replied:
Applied the suggested fix. Thanks!

guggero reviewed (approved):
Re-ACK.
